packages/css-selector-parser/src/ast-tools/specificity.ts

```diff
@@ -0,0 +1,105 @@
+import type {
+  ImmutableSelectorNode,
+  ImmutablePseudoClass,
+  SelectorNode,
+} from "../ast-types";
+import { walk } from "./walk";
+
+type Specificity = [number, number, number, number];
+export function calcSpecificity(ast: ImmutableSelectorNode): Specificity {
+  const result: Specificity = [0, 0, 0, 0];
+  // ToDo: remove casting once immutable walk is supported
+  walk(ast as SelectorNode, (node) => {
+    switch (node.type) {
+      case `type`:
+      case `pseudo_element`:
+        result[3]++;
+        break;
+      case `class`:
+      case `attribute`:
+        result[2]++;
+        break;
+      case `pseudo_class`:
+        if (customPseudoClass[node.value]) {
+          customPseudoClass[node.value](node, result);
+          return walk.skipNested;
+        }
+        result[2]++;
+        break;
+      case `id`:
+        result[1]++;
+        break;
+    }
+    return node.type !== `selector` && node.type !== `compound_selector`
+      ? walk.skipNested
+      : undefined;
+  });
+  return result;
+}
+
+const customPseudoClass: Record<
+  string,
+  (node: ImmutablePseudoClass, result: Specificity) => void
+> = {
+  not: mostSpecificInnerSelector,
+  is: mostSpecificInnerSelector,
+  has: mostSpecificInnerSelector,
+  where: () => {
+    /* no specificity*/
+  },
+  "nth-child": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-last-child": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-of-type": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-last-of-type": pseudoClassPlusMostSpecificInnerSelector,
+};
+
+function pseudoClassPlusMostSpecificInnerSelector(
+  node: ImmutablePseudoClass,
+  result: Specificity
+) {
+  result[2]++;
+  mostSpecificInnerSelector(node, result);
+}
+function mostSpecificInnerSelector(
+  node: ImmutablePseudoClass,
+  result: Specificity
+) {
+  if (node.nodes) {
+    let highest: Specificity | undefined;
+    for (const selector of node.nodes) {
+      const currentSpecificity = calcSpecificity(selector);
+      if (!highest || compareSpecificity(currentSpecificity, highest) === 1) {
+        highest = currentSpecificity;
+      }
+    }
+    addSpecificity(result, highest!);
+  }
+}
+/**
+ * compare 2 specificities
+ * @param a first specificity
+ * @param b second specificity
+ * @returns 0 if equal, 1 when a is more specific, -1 when b is more specific
+ */
+export function compareSpecificity(a: Specificity, b: Specificity): -1 | 0 | 1 {
```

Comment:
maybe do the same as we did in the walk....add object with these value meanings

{lower: -1, same: 0, higher: 1}

Reply:
unlike the walk, this is internal, and I think between this being small and with jsdocs it doesn;t need any addition

Reply:
no its not.....it is exported....but ok maybe not